### DISCUSSION TITLE
HiddenField is not called for patch verb

### DISCUSSION THREAD

**Discussion by abtinmo** [encode/django-rest-framework#8259] [Question & Answer]
class ClubValidateSerializer(serializers.ModelSerializer):
        user = serializers.HiddenField(default=serializers.CurrentUserDefault())

      class Meta:
          model = Club
          fields = ('id', 'validated', 'user')
          read_only_fields = ('id',)
  
      def update(self, instance, validated_data):
          if instance.validated == validated_data['validated']:
              return instance
          history = ClubValidateHistory(club=instance, validator=validated_data['user'],
                                        from_state=instance.validated, to_state=validated_data['validated'])
          instance.validated = validated_data['validated']
          instance.save()
          history.save()
          return instance

Expected behavior

I want to get the user field in validated_data, but it throws KeyError. It from a patch update.
Actual behavior

history = ClubValidateHistory(club=instance, validator=validated_data['user'],
KeyError: 'user'
"PATCH /backend/clubs/4/validate/ HTTP/1.1" 500 17633

link to original issue https://github.com/encode/django-rest-framework/issues/6560

**Comment by denisSurkov** [comment]
I have the same issue:

## The problem

Docs [state](https://github.com/encode/django-rest-framework/blob/7658ffd71d4ad07ddada20b2b9538b889ec02403/docs/api-guide/fields.md?plain=1#L541C4-L541C4):


> ### HiddenField
> ...
> For example, to include a field that **always** provides the current time as part of the serializer validated data, you would use the following:

But HiddenField is skipped during validation if serializer has `partial=True`. So HiddenField does **not** always provide default value.



## How to reproduce
```python3
from rest_framework import serializers


class DemoCallable:
    def __call__(self, *args, **kwargs):
        return 'does not matter'


class DemoSerializerWithHiddenField(serializers.Serializer):
    some_field = serializers.CharField()
    demo_hidden = serializers.HiddenField(default=DemoCallable())

    def validate(self, attrs):
        print(attrs)
        return attrs


partial_serializer = DemoSerializerWithHiddenField(data={
    "some_field": "hidden ignored"
}, partial=True)
partial_serializer.is_valid(raise_exception=True)

print(partial_serializer.validated_data)
# OrderedDict([('some_field', 'hidden ignored')])
# OrderedDict([('some_field', 'hidden ignored')])


serializer = DemoSerializerWithHiddenField(data={
    "some_field": "hidden not ignored"
}, partial=False)
serializer.is_valid(raise_exception=True)

print(serializer.validated_data)
# OrderedDict([('some_field', 'hidden not ignored'), ('demo_hidden', 'does not matter')])
# OrderedDict([('some_field', 'hidden not ignored'), ('demo_hidden', 'does not matter')])
```

## Possible solution
```python3
from rest_framework import serializers

class HiddenField(serializers.HiddenField):
    def validate_empty_values(self, data):
        return (True, self.get_default())

    def get_default(self):
        return self.default(self)


class DemoCallable:
    def __call__(self, *args, **kwargs):
        return 'does not matter'


class DemoSerializerWithCustomHiddenField(serializers.Serializer):
    some_field = serializers.CharField()
    demo_hidden = HiddenField(default=DemoCallable())

    def validate(self, attrs):
        print(attrs)
        return attrs

partial_serializer = DemoSerializerWithCustomHiddenField(data={
    "some_field": "custom hidden not ignored"
}, partial=True)
partial_serializer.is_valid(raise_exception=True)

print(partial_serializer.validated_data)
# OrderedDict([('some_field', 'hidden not ignored'), ('demo_hidden', 'does not matter')])
# OrderedDict([('some_field', 'hidden not ignored'), ('demo_hidden', 'does not matter')])


serializer = DemoSerializerWithCustomHiddenField(data={
    "some_field": "custom hidden not ignored also"
}, partial=False)
serializer.is_valid(raise_exception=True)

print(serializer.validated_data)
# OrderedDict([('some_field', 'custom hidden not ignored also'), ('demo_hidden', 'does not matter')])
# OrderedDict([('some_field', 'custom hidden not ignored also'), ('demo_hidden', 'does not matter')])
```

## Caveats

I think it's irresponsible to ignore overriding solution `HiddenField` breaks `partial` behavior, bacause `partial` has to return only provided subset. 

I see two options:
- Mention that `HiddenField` does not appear on partial (aka PATCH) and don't touch `HiddenField` implementation;
- Fix `HiddenField` but change version. I don't like the idea changing major version, but i don't know how to do it other way.

**Reply by denisSurkov** [reply]
@auvipy, sorry to bother you, could you help with this discussion? Or mention someone, who might help with this one.

**Reply by auvipy** [reply]
IMHO we can start with the option 1. and if the current implementation is buggy, we need to fix that in a major release :)

**Reply by denisSurkov** [reply]
but how to know "if the current implementation is buggy"? :)

it's like

>  ### HiddenField
>
>    ...
>   For example, to include a field that always provides the current time as part of the serializer validated data, you would use the following:
>
>  #### Caution
> HiddenField does not appear in partial serializer when making PATCH requests.

And what a developer should do if it's important to include hidden field in patch requests?

Maybe we could add `is_partial=False` flag to HiddenField kwargs.

**Reply by denisSurkov** [reply]
@auvipy , can I add this "Caution" notice, then we would continue discussing is this behavior okey or not?

I'll make PR

**Reply by auvipy** [reply]
yes please

**Reply by auvipy** [reply]
I will check the code that time as well

**Reply by denisSurkov** [reply]
@auvipy , are there any updates?

**Reply by auvipy** [reply]
I'm bit busy with other priorities right now.
